Part1_InterruptControllersEvolution.html

```diff
@@ -104,79 +104,78 @@ <h3>PIC</h3>
+	<p>The last variant of APIC was good, but not without downsides. All of the interrupt lines from devices made the system very complicated and thus increased error probability. The PCI express bus came to replace the PCI bus, which simplified all interrupt systems completely. It doesn't have interrupt lines at all. For backwards compatibility interrupt signals (INTx#) are emulated with a separate kind of messages. With PCI interrupt lines their connection was made with physical wires. With PCI express interrupt lines a connection is logical and is made by PCI express bridges. But this support of legacy INTx interrupts only exists for backwards compatibility with the PCI bus. PCI express introduces a completely new method of interrupt delivery - MSI (Message Signaled Interrupts). In this method a device signals about the interrupt simply by writing to a special place in the MMIO region of the CPUs LAPIC.<p/>
 
-	<p>Earlier one PCI device (this means all its functions) could have only 4 interrupts, but now it is became possible to address up to 32 interrupts.<p/>
+	<p>Earlier a single PCI device (this means all its functions) could have only 4 interrupts, but now it became possible to address up to 32 interrupts.<p/>
 
-	<p>In a case of MSI there is no sharing for interrupt lines, every interrupt naturally corresponds to its device.<p/>
+	<p>In the case of MSI there is no sharing of interrupt lines: every interrupt naturally corresponds to its device.<p/>
 
-	<p>MSI interrupts also solve one more problem. For example let's imagine a situation where a device makes a memory-write transaction, and wants to signal about its completion through the interrupt. But a write transaction can be delayed on the bus in a process of its transmission (and the device couldn't know about it). In this case a signal about the interrupt will come to CPU first, so a processor will read yet not valid data. If MSI is used, information about MSI is transmitted in the same way as data messages, and it can't came earlier.<p/>
+	<p>MSI interrupts also solve one more problem. For example, let's imagine a situation where a device makes a memory-write transaction, and wants to signal about its completion through the interrupt. But a write transaction can be delayed on the bus in the process of its transmission (and the device couldn't know about it). In this case the signal about the interrupt will come to the CPU first, so the processor will read invalid data. If MSI is used, information about the MSI is transmitted in the same way as data messages, and so it can't come earlier.<p/>

```

Comment:
Is it incorrect to say "will read yet not valid data"? Cause I think it has more meaning than just "will read invalid data"

Reply:
Makes sense. In that case "not yet" should be used

Reply:
```suggestion
	<p>MSI interrupts also solve one more problem. For example, let's imagine a situation where a device makes a memory-write transaction, and wants to signal about its completion through the interrupt. But a write transaction can be delayed on the bus in the process of its transmission (and the device couldn't know about it). In this case the signal about the interrupt will come to the CPU first, so the processor will read yet not valid data. If MSI is used, information about the MSI is transmitted in the same way as data messages, and so it can't come earlier.<p/>
```

Reply:
"not yet" is more natural than "yet not"

Reply:
```suggestion
	<p>MSI interrupts also solve one more problem. For example, let's imagine a situation where a device makes a memory-write transaction, and wants to signal about its completion through the interrupt. But a write transaction can be delayed on the bus in the process of its transmission (and the device couldn't know about it). In this case the signal about the interrupt will come to the CPU first, so the processor will read not yet valid data. If MSI is used, information about the MSI is transmitted in the same way as data messages, and so it can't come earlier.<p/>
```